### PR TITLE
[CV2-3402] - make audio always similarity

### PR DESCRIPTION
**Pull request by computermacgyver**
This PR includes:
*  updated tests and made a few further tweaks to ensure we always treat audio as similarity throughout the whole stack.
* cleaned up the perl similarity function functions and added the reference implementations to the repo at extra/audio_similarity (These are previously in a private repo called research-audiosimilarity)
* Edits to allow the Perl functions to pass “strict” mode. I also set the min_overlap to be equal to span. We set span at 20 but then adjust it to be equal to the length of the shortest file. So, for files longer than 20 seconds behavior is unchanged. For short files, we require the whole file to match the other.
* Added a reference implementation of the perl similarity functions for testing THIS CODE IS NOT USED IN PRODUCTION

Additional context:
Around 8 months ago, there was a change that flipped it to distance (zero is best), but only in the main case: 
When we introduced code to not even compare very different audio, we hardcoded a return value of zero: 
 , that is at odds with the rest of the function after it was flipped.

Reference: CV2-3402

### CODE REVIEW

**Comment by computermacgyver** [comment]
Marking this as a draft. We also need to run Check Web full stack tests

**Comment by DGaffney** [comment]
Lets push this function definition into QA and then re-run these examples that you ran through @computermacgyver? If we can confirm true positives and true negatives then I'm satisfied. FWIW, we can re-define that function on the fly so don't even need to merge this on QA to test this...

File: extra/audio_similarity/python_scorer.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
+import numpy as np
```
[the remark this answers is not in the document]
Comment (computermacgyver) [reply extra/audio_similarity/python_scorer.py:1]: Not going to add to requirements.txt just for a reference implementation

File: extra/audio_similarity/python_scorer.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        listx = listx[:len(listy)]
+    if min(len(listx), len(listy)) < min_overlap:
+        # Error checking in main program should prevent us from ever being
+        # able to get here.
+        return 
+    #raise Exception('Overlap too small: %i' % min(len(listx), len(listy)))
+    return correlation(listx, listy)
+  
+# cross correlate listx and listy with offsets from -span to span
+def compare(listx, listy, span, step):
```
[the remark this answers is not in the document]
Comment (computermacgyver) [reply extra/audio_similarity/python_scorer.py:39]: If only Python had proper variable scoping!

**Comment by computermacgyver** [comment]
Note integration tests pass here - https://github.com/meedan/check-web/pull/1517

File: extra/audio_similarity/python_scorer.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+#listy=[30,40,50,60,70,80]*5;
+#listx=[1,2,3,4,5,6,7,8,9];
+#listy=[4,5,6];
+#print(get_score(listx,listy))
+#print(correlation(listx,listy))
+
+#for offset in range(-2,3,1):
+#	print(cross_correlation(listx,listy,offset))
+#print(compare(listx,listy,150,1));
+print(get_score(arg1, arg2))
```
[the remark this answers is not in the document]
Comment (skyemeedan) [reply extra/audio_similarity/python_scorer.py:83]: OH, I just understood that this is all testing code!

File: manage.py [hunk 4]
```diff
@@ -122,6 +122,9 @@ def init_perl_functions():
         CREATE OR REPLACE LANGUAGE plperl;
       """)
     )
+    # DO NOT EDIT HERE.
```
Comment (skyemeedan) [inline manage.py:125]: Maybe consider moving this into app/main/lib/shared_models/audio_model.py (or adjacent file?) and just importing into manage.py to run the installation?  Probably would make easier to update if all the code needing updates is in the same place, and that way the actual algorithm logic is not in the control script

**Review by skyemeedan — COMMENTED** [review]
in the ticket it says:

File: extra/audio_similarity/python_scorer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (computermacgyver) [reply extra/audio_similarity/python_scorer.py:83]: Yeah, this is the reference implementation. This code was written before I ever tried to write it in Perl :joy:

**Comment by computermacgyver** [comment]
> whoops, accidentally clicked submit too soon on the review.
> 
> > The files used in the Check Web integration tests have fingerprints that are only 19 units long (one unit is roughly one second).
> > ...
> > We have a few options here.
> > Update the Check Web tests to use longer files.
> > Decrease the minimum overlap required from 20 to something like 15 or 10 units
> > Adjust minimum overlap to be a percentage of the shorter file (e.g., 90%)
> 
> I didn't notice if that was addressed here yet, do we need another ticket for that?

Yes, this is addressed. I've tweaked the minimum overlap required so that it is whatever is less of 20 seconds (our previous fixed value) or the length of the shorter clip

File: manage.py (hunk 4 above)
Comment (computermacgyver) [reply manage.py:125]: manage.py is where all our database functions live.

File: extra/audio_similarity/python_scorer.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+def cross_correlation(listx, listy, offset, min_overlap):
+    if offset > 0:
+        listx = listx[offset:]
+        listy = listy[:len(listx)]
+    elif offset < 0:
+        offset = -offset
+        listy = listy[offset:]
+        listx = listx[:len(listy)]
+    if min(len(listx), len(listy)) < min_overlap:
+        # Error checking in main program should prevent us from ever being
```
[the remark this answers is not in the document]
Comment (computermacgyver) [reply extra/audio_similarity/python_scorer.py:32]: This is the reference implementation that is not actually used. In the Perl implementation that we use within Postgres there is not the equivalent of an assert.

File: manage.py [hunk 6]
```diff
@@ -199,6 +204,9 @@ def init_perl_functions():
         LANGUAGE plperl;
       """)
     )
+    # DO NOT EDIT HERE.
+    # Please see the reference implementations at /extra/audio_simliarity
+    # Edit those implementations and ensure output is correct for the test data there.
     sqlalchemy.event.listen(
       db.metadata,
```
[the remark this answers is not in the document]
Comment (computermacgyver) [reply manage.py:212]: Just added. Thanks.

File: extra/audio_similarity/python_scorer.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+def compare(listx, listy, span, step):
+    if span > min(len(listx), len(listy)):
+        span = min(len(listx), len(listy)) -1
+    min_overlap = span
+    corr_xy = []
+    for offset in np.arange(-span, span + 1, step):
+        #print(offset);
+        corr_xy.append(cross_correlation(listx, listy, offset, min_overlap))
+    return corr_xy
+
```
[the remark this answers is not in the document]
Comment (computermacgyver) [reply extra/audio_similarity/python_scorer.py]: Thanks. Done

**Review by skyemeedan — APPROVED** [review]

**Review by DGaffney — APPROVED** [review]
Great work here. Lets give it a spin!

**Pull request merged**